### ISSUE TITLE
0.2.14 Caused a breaking change

### ISSUE THREAD

**Issue by OverZealous** [angular-ui/ui-router#1915]
I (accidentally) just upgraded to the latest release, and suddenly my tests won't pass with this error:

```plain
 Error: Default value (false) for parameter 'error' is not an instance of Type (any)
          at $$getDefaultValue (./vendor/angular-ui-router/release/angular-ui-router.js:1628)
          at $value (./vendor/angular-ui-router/release/angular-ui-router.js:1643)
          at ./vendor/angular-ui-router/release/angular-ui-router.js:1686
          at forEach (./vendor/angular/angular.js:323)
          at ./vendor/angular-ui-router/release/angular-ui-router.js:1687
          at transitionTo (./vendor/angular-ui-router/release/angular-ui-router.js:3148)
          at go (./vendor/angular-ui-router/release/angular-ui-router.js:3069)
          at ./build/connect/js/app/routes/error/error.route.js:5:0 <- null:0:0
          at check (./vendor/angular-ui-router/release/angular-ui-router.js:2008)
          at update (./vendor/angular-ui-router/release/angular-ui-router.js:2020)
          at ./vendor/angular/angular.js:14712
          at ./src/connect/app/routes/routes.spec.js:30
```

I don't mind fixing the issue (it worked without error on 0.2.13), but this is a breaking change on a what should be a patch release.

The faulty code in question looks like this:
```js
$stateProvider.state('error', {
    params: {
        error: false,
        fromState: {},
        fromParams: {},
        toState: {},
        toParams: {},
    },
   //...
```

-----

Actually, on reading through the docs again, this seems like it should be an acceptable format, so this might actually be a legitimate bug.

**Comment by nateabele** [comment]
Yeah, seems like it might be a bug. I'll investigate shortly.

**Comment by eddiemonge** [comment]
```
.config ( $stateProvider ) ->
  $stateProvider

  .state 'packages',
    abstract: true
    parent: 'layout'
    template: '<ui-view />'
    url: '/packages/'

  .state 'packages.received',
    url: 'received?page&per_page&archived&q'
    reloadOnSearch: false
    params: 
      page:
        squash: true
        value: 1
      per_page:
        squash: true
        value: 10
      archived:
        squash: true
        value: false
      q: squash: true
    templateUrl: '/modules/packages/index.html'
    controller: 'PackagesIndexController'
     
```

*sidenote* squash never seemed to squash the query string part even when it matched. im guessing its because of type as well

**Comment by compact** [comment]
I got the same error after updating from 0.2.13 to 0.2.14. My state parameter is also a boolean that defaults to false, if that's relevant.

**Comment by philBrown** [comment]
I'd say the boolean param property is definitely causing the error.

Removing any boolean `param` properties in state config makes the error go away.

**Comment by lazar-vuckovic** [comment]
Same error here. Only shows up when setting the default parameter to "false".

**Comment by OtaK** [comment]
:+1: Needs to be fixed ASAP. It took down our production momentarily.

**Comment by zacaj** [comment]
seems to be https://github.com/angular-ui/ui-router/commit/da63b27b7bbd13d2633b68fa74606f977e6d4147#diff-6de0577dce2b7e2e060fb6f1b36a33e3R1628, when self.type is 'any', self.type.is(x) simply returns x, and when x is false, or "" in my case, !self.type.is() is throwing the error

**Comment by spencerwi** [comment]
This affected our production as well. It happened when the default parameters were `false` or `0`, leading me to believe that this is a bad falsiness check.

**Comment by elado** [comment]
+1

Default value `false` for a param (`params: { blah: false }`) throws 

```
Error: Default value (false) for parameter 'blah' is not an instance of Type (any)
```

**Comment by spencerwi** [comment]
I'm not one to be the semver police, but introducing breakages for something like this in a `_._.x` release strikes me as a dangerous move -- our production instance got hit by this because anyone who (like us) ran `bower install ui-router --save` when 0.2.13 was the newest version gets from bower a dependency on `~0.2.13` -- which will grab 0.2.14. Meaning that without any code changes, two consecutive builds may go from "working" to "broken".

The simple solution is to manually ensure that apps depend on *exactly* 0.2.13, but it takes debugging down (or googling and finding this GH issue) to figure out that's the fix.

**Comment by nateabele** [comment]
>  I'm not one to be the semver police, but introducing breakages for something like this [...]

@spencerwi Yup, trust me, didn't do it on purpose. :wink:

**Comment by nateabele** [comment]
> The simple solution is to manually ensure that apps depend on exactly 0.2.13, but it takes debugging down (or googling and finding this GH issue) to figure out that's the fix.

Another possible solution might be to help us help you by also running your builds against UI-Router HEAD.

**Comment by OtaK** [comment]
Apart from the debate that "a commit broke everything, blah blah", maybe we should also consider URI standards. URI querystrings values *must* be strings only as per HTTP spec.
As such, standards enforcement like this is more beneficial than not. Falsey values don't get appended to query string, whereas truthy (ints/strings) ones get actually appended if they are actually strings. (it shouldn't break on ints though)

Still, I guess you're running unit tests on routing as well here, and the build should not have been released without the tests passing. Maybe you can remove the 0.2.14 release on bower/npm/whatever no as an emergency measure to prevent people breaking their builds from now on until you solve the problem?

**Comment by spencerwi** [comment]
@nateabele If I had a personal Angular project, mosdef. Company product needs to stay pretty stable, though. 

OtaK raises a good point: unit tests. That might be a good alternative way to dive in and help.

Big fan of the work you've done already -- UI-Router is *the* de-facto router for Angular, to the point that the Angular core team itself is basing from it. That's no small feat.

**Comment by philBrown** [comment]
@OtaK 

> maybe we should also consider URI standards. URI querystrings values must be strings only as per HTTP spec

The thing is, the `stateConfig.params` map may contain...

> additional **non-url** parameters

From http://angular-ui.github.io/ui-router/site/#/api/ui.router.state.$stateProvider#methods_state

**Comment by nateabele** [comment]
> Still, I guess you're running unit tests on routing as well here, and the build should not have been released without the tests passing.

@OtaK https://travis-ci.org/angular-ui/ui-router/branches

We run a suite of ~800 assertions against 4 different versions of Angular _on every commit_, so I have no idea what you're talking about.

If there are other OS projects you use whose test suites have a perfect track record of preventing regressions, let me know, because I'd love to learn their secrets. :wink:

> @nateabele If I had a personal Angular project, mosdef. Company product needs to stay pretty stable, though.

@spencerwi Sorry for not being clear. I meant running multiple builds in parallel. We do it for our products / client projects on any OS projects we use in a high-visibility capacity (especially JS crap, because you just never know). Decent build tools make this trivial.

If there's any way we can make it easier for people to run their production apps' test suites against master (again, just the test suites, not actually deploying it to production), let us know, because that's the best thing we can possibly do to prevent regressions.

> OtaK raises a good point: unit tests. That might be a good alternative way to dive in and help.

Absolutely. If we push a change that breaks your apps, the best response we can get is a PR with a unit test. Again, if there's anything we can do to make this easier, let us know.

**Comment by nateabele** [comment]
@spencerwi Thanks for your kind words, btw.

**Comment by inca** [comment]
Also breaks when default param value is integer `0` (which also happens to be falsey).

**Comment by inca** [comment]
@OtaK BTW (and a little off the topic), there's nothing wrong with query string parameter values like `0`, `false` and `null` which can be parsed from strings server-side.

**Comment by shawndewet** [comment]
I find it fascinating that there are complaints about a v0.x release breaking PRODUCTION code!! Guys, what is a PRODUCTION system doing with references to a v0.x library?  Once it's in v1.x, complain away.  In the meantime, @nateabele and crowd, keep up the good work...amazing though that ~800 tests didn't pick up this issue.

**Comment by inca** [comment]
@shawndewet semantic versioning states that minor releases should be treated as breaking in 0.x.x while patch releases shouldn't, that's all. There are millions of 0.x.x libraries doing well in production (Node.js itself).

**Comment by JaKXz** [comment]
I've [tried passing a params object to `$state.go`](https://github.com/angular-ui/ui-router/issues/928#issuecomment-101079545) with versions `0.2.12`, `0.2.13`, and `0.2.14` but my [`$stateParams` object is *not* being populated](https://github.com/angular-ui/ui-router/blob/master/src/state.js#L885).

I've made some noise about this today on that closed issue, sorry about that - Google took me to those issues before this one - would really love some help though

**Comment by inca** [comment]
@JaKXz your state should include `params: {}` section where you specify which parameters state accepts (and their default values), like this:

```js
$stateProvider.state('some', {
  params: {
    foo: 'bar',
    another: 1
  }
});
```

**Comment by JaKXz** [comment]
@inca thanks for the response - but unfortunately passing the object is always returning the default value:
```javascript
$stateProvider.state('some', {
  params: {
    foo: {}
  }
});
```
even though I'm definitely passing it a valid object (as described before).

**Comment by JaKXz** [comment]
Turns out I had a `$state.go` call in my `$stateChangeStart` listener which wasn't passing on the `toParams`. Thanks for the help @inca!

**Comment by eddiemonge** [comment]
This seems to have been fixed in https://github.com/angular-ui/ui-router/commit/ba6ed61ba10627607b32e5a0cef030bed3cb58c5 if people want to test and report back on it

**Comment by jodytate** [comment]
i can confirm that ba6ed61 fixes this problem for our code

**Comment by nateabele** [comment]
@OverZealous Can you confirm the same?

**Comment by OverZealous** [comment]
I can't check it right now, but I'm confident this will fix it.  Thanks!

**Comment by christopherthielen** [comment]
Closed via ba6ed61ba10627607b32e5a0cef030bed3cb58c5

**Comment by levivm** [comment]
The bug persists to me. 

I'm using 0.2.14 and when I try `params: { foo: false }) ` it throws an error:

`Error: Default value (false) for parameter 'foo' is not an instance of Type (any)`
